### ISSUE TITLE
Heads up: large incompatible V8 API changes coming up

### ISSUE THREAD

**Issue by bnoordhuis** [nodejs/nan#289]
https://code.google.com/p/v8/issues/detail?id=3929

https://groups.google.com/forum/#!topic/v8-users/gQVpp1HmbqM - outlines the changes.

Some of the changes already landed upstream and will land in io.js in two or three months.

**Comment by kkoopa** [comment]
Oh my. Again?

**Comment by kkoopa** [comment]
I wonder if these maybe monads can be backported to the old API somehow. It would probably be the nicer option.

**Comment by kkoopa** [comment]
kkoopa/nan@c66082dee22375c65cbd2563a4457cf7fb7eeac5

**Comment by kkoopa** [comment]
Made a small list of major changes so far in v8 4.3, it is not exhaustive. All of these have been adapted to return `Maybe`s.
```c++
Script::Compile
Script::Run
Script::CompileUnbound

Value::Converters
Value::Equals

String Factory Methods

Object::Mostly Everything

Array::CloneElementAt

Function::NewInstance
Function::Call

Date::New

RegExp::New

FunctionTemplate::GetFunction

ObjectTemplate::NewInstance

TryCatch::StackTrace
```
and then IdleNotification was renamed and changed its argument `bool IdleNotification(int idle_time_in_ms) --> bool IdleNotificationDeadline(double deadline_in_seconds)`

**Comment by kkoopa** [comment]
Having looked at the necessary changes as of latest 4.3, it does not make sense to backport the `Maybe` types. It is easier to just unwrap the necessary return values for new versions that use maybe types internally. No v8 functions seem to accept `MaybeLocal<T>` or `Maybe<T>` as arguments, so the whole system seems quite broken. You always end up wanting to unwrap.

**Comment by domenic** [comment]
> You always end up wanting to unwrap.

Sure, but you want to unwrap and decide for yourself how to handle the None case, right? Instead of #310, which just calls `.ToLocalChecked()` always.

There are a number of issues opened on Chromium where incorrect `.ToLocalChecked()` changes cause crashes. It seems better if nan does not automatically encourage this buggy style.

**Comment by kkoopa** [comment]
I thought of emulating old behavior, not doing a check in the conversion, essentially getting rid of the new API altogether by making it conform to the old. But, if this is too restrictive, the other alternative is backporting the new API, as I originally did https://github.com/kkoopa/nan/commit/c66082dee22375c65cbd2563a4457cf7fb7eeac5.

In that case, it might make also make sense to introduce a bind operator `>>=` for chaining functions that output Maybes. The major problem is getting rid of as much undue clumsiness as possible.

```haskell
(>>=) :: Maybe a -> (a -> Maybe b) -> Maybe b
Nothing  >>= _  =  Nothing    -- A failed computation returns Nothing
(Just x) >>= f  =  f x        -- Applies function f to value x
```

**Comment by domenic** [comment]
Hmm yeah https://github.com/kkoopa/nan/commit/c66082dee22375c65cbd2563a4457cf7fb7eeac5 seems like at least you aren't inviting errors. It'd be cool if there were a way to avoid the extra "Maybe", e.g. `namespace n = nan::maybe` or something, since otherwise you "punish" people who do the right thing with an extra `n`.

And sure, adding whatever helpers you'd use seems fine.

**Comment by kkoopa** [comment]
What do you mean by avoiding the extra "Maybe"? What is extra where?
A further complication is that some constructors now return a `MaybeLocal`, whereas others return a `Local` directly, e.g. `String::FromUtf8Value` returns a `MaybeLocal<String>`, whereas `String::Empty()` returns a `Local<String>`. As both go through `NanNew<String>()`, it would likely be necessary to harmonize by always returning a `MaybeLocal`, even when an empty `String` is being constructed.

**Comment by domenic** [comment]
> What do you mean by avoiding the extra "Maybe"? What is extra where?

I mean the extra at the call site in each called method name. Client code ends up more verbose if it wants to do the right thing and handle errors. (But then again I guess the actual error-handling code will make it more verbose anyway, in most cases, so maybe this isn't a big deal.)

**Comment by kkoopa** [comment]
I still don't quite follow. For example, let's take `NanBooleanValue`, which wraps `Value::BooleanValue`. It is defined as:
```c++
NAN_INLINE NanMaybe<bool> NanBooleanValue(v8::Local<v8::Value> val) {
  return val->BooleanValue(NanGetCurrentContext());
}
```
for v8 >= 4.3 and for lesser versions it is defined as
```c++
NAN_INLINE NanMaybe<bool> NanBooleanValue(v8::Local<v8::Value> val) {
  return NanMaybe<bool>(val->BooleanValue());
}
```
It would be used as
```c++
NAN_METHOD(BooleanValue) {
  NanScope();
  NanReturnValue(NanBooleanValue(args[0]).ToValueChecked());
}
```

Are you referring to the `return NanMaybe(foo)` in the definition of `NanBooleanValue` for versions less than 4.3? I don't see how a namespace alias helps, as NAN does not (yet) make proper use of namespaces, something which could be fixed to some extent in NAN 2.0, aside from the macros, which cannot be namespaced.

**Comment by domenic** [comment]
I'm sorry, I totally misinterpreted https://github.com/kkoopa/nan/commit/c66082dee22375c65cbd2563a4457cf7fb7eeac5. You are right!

**Comment by kkoopa** [comment]
@agnat Is it possible to break up the `NanNew` factories so they can return `Local<T>` for some arguments to `NanNew<T>` and `NanMaybeLocal<T>` for other args?

```c++
Local<String> s = NanNew<String>();
NanMaybeLocal<String> ss = NanNew<String>("foo");
```

Would this be confusing? Is it better to stick to one single return type? thereby having `NanMaybeLocal<String> s = NanNew<String>()`

**Comment by agnat** [comment]
> Is it possible to break up the NanNew factories so they can return Local<T> for some arguments to NanNew<T> and NanMaybeLocal<T> for other args?

Nope. Not without considerable effort and a lot more angular bracket mumbo-jumbo.

> Would this be confusing?

Yes.

> Is it better to stick to one single return type?

Hm, probably... but I don't like the consequences of that either.

This is a nasty one... :-/

**Comment by kkoopa** [comment]
As of now, I think `NanNew<T>` should always return a `NanMaybeLocal<T>` if `T` is `Script`, `Date` or `RegExp`, or if `T` is `String` and no arguments are given. Otherwise, it should return a `Local<T>`.

Maybe the only sort of acceptable solution would be introducing `NanMaybeLocal<T> NanNewMaybe<T>(...)` for `String` with args, `Script`, `Date` and `RegExp`, and removing them from `NanNew<T>`? If so, what's a better name than NanNewMaybe?

I previously speculated about introducing a monadic bind operator, but I'm afraid it cannot be done. I really don't like this change in v8, it makes everything so clumsy.

**Comment by agnat** [comment]
> As of now, I think NanNew<T> should always return a NanMaybeLocal<T> if T is Script, Date or RegExp

I _think_ you can switch those to return maybe-foo  by modifying the return type typedef in the factories. Just to see how varying/mixed return types would feel.

String needs a bit more doing, though. But if it's just the one... let me think about it.

**Comment by kkoopa** [comment]
Yeah, they should work okay, it's the empty string constructor that's the snag. At one point I tried something like
```c++
template <typename T> struct MaybeFactoryBase { typedef NanMaybeLocal<T> return_t; };

template <>
struct Factory<v8::Date> : MaybeFactoryBase<v8::Date> {
  static inline return_t New(double value);
};
```
And it worked up until the `String`.

**Comment by kkoopa** [comment]
Solving that should require either an overload or breaking the empty `String` constructor out into its own function entirely.

On a side note, this will be a breaking change in NAN no matter what, so that means NAN 2.0 can clean up a lot of old cruft and mistakes, e.g. #219.

**Comment by kkoopa** [comment]
Fixed in e0200b80d7626c7747d5e2c430ce07629a463480
